### PR TITLE
feat: kanban view reimplementation

### PR DESCRIPTION
**Pull request by wingkwong**
## Change Summary

ref: #140 #3716

- [x]  Meta DB Migration Script
  - [x]  two new columns in `nc_kanban_view_v2` 
    - [x]  `grp_column_id`: grouping field column id (used in stackedBy)
    - [x]  `meta`: meta data like stack order, colors, option titles etc.
- [x]  stackedBy in Toolbar
  - [x] dropdown for selecting a Single Select column
  - [x] rerendering kanban view based on the change
- [x]  Kanban View
  - [x]  Basic View Operations
    - [x] Create View
      - [x] Select the first grouping field
    - [x] Copy View
    - [x] Rename View  
    - [x] Delete View
  - [x]  Stack 
    - [x]  Create a new Stack 
    - [x]  Re-order Stack
    - [x]  Add records within a Stack
      - [x] Preset grouping field value
    - [x]  Drag-and-drop record cards to other Stack
    - [x] Scroll to fetch records
    - [x] Dropdown
      - [x]  Add new record
      - [x] Collapse Stack
      - [x] Delete Stack
  - [x]  Record Card
    - [x] Click to expand form to edit
    - [x]  Drag-and-drop records across stacks
    - [x] Content Menu
      - [x] Expand Record
      - [x] Delete Record
  - [x] Share View 
    - [x] With Password Protected
    - [x] With Download Allowed
  - [x] Misc
    - [x] Sync grouping field options from other views (e.g. remove options in grid view)
    - [x] Link to Row Integration
    - [x] Prevent Grouping Field Column Deletion
    - [x] Grouped APIs
      - [x] dbTableRow: /api/v1/db/data/{orgs}/{projectName}/{tableName}/group/{columnId}
      - [x] dbViewRow: /api/v1/db/data/{orgs}/{projectName}/{tableName}/views/{viewName}/group/{columnId}
      - [x] public: /api/v1/db/public/shared-view/{sharedViewUuid}/group/{columnId}
  - [ ] Cypress
  - [x] Crowdin
  - [x] v-e

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Test scenario's
: Using sakila DB, Film table

- [x] Create kanban 
- [x] Rename kanban
- [x] Duplicate kanban
- [x] Configure grouping field: `Rating`
  - [x] Verify stack count
  - [x] Verify stack order
  - [x] Verify footer records count
  - [x] Verify per stack card count
  - [x] Verify card order within a stack
- [x] Toolbar menu 
  - [x] Hide fields
  - [x] Enable filter
  - [x] Disable filter
  - [x] Sort fields
  - [x] Reset sort
- [x] Drag drop
  - [x] Inter stack card drag-drop
  - [x] Intra stack card drag-drop [will be reset if we re-open kanban]
  - [x] Stack drag-drop reorder
- [x] Stack context menu
  - [x] Add record
  - [x] Rename stack
  - [x] Delete stack
  - [x] Collapse stack
- [x] Expand record
  - [x] Fields visible
  - [ ] Update record  
- [x] ~~View context menu~~
  - [x] ~~Download CSV~~
  - [x] ~~Upload CSV~~
  - [x] ~~Locked view~~
- [ ] Share view
- [ ] Kanban stack handling
  - [ ] Change grouping field
  - [ ] Add / edit stack
    - [x] Add option
    - [ ] Rename option
    - [ ] Delete option 
- [ ] Options & menu accessibility for different roles 
- [x] Delete kanban

## Screenshot

image

### CODE REVIEW

**Comment by dstala** [comment]
Some observations or comments based on a dry run; some of the items listed here might be WIP

- [ ] Pick the first available `select field` by default on creating a new Kanban view
- [ ] Stacked by `Field name`: if easy, display `Field name` in bold & colored
- [ ] [airtable] Display on `Primary Value` by default on the card. Hide the rest of the fields
  - [ ] [airtable] Consider displaying only `Primary Value`; ignore column type if it's any easy change
- [ ] Hide the `select field` context menu on the card in the Locked view 
- [ ] For Kanban, often we will end up hiding many fields. 
  - This will lead issue to insert a new record (https://github.com/nocodb/nocodb/issues/3513)
  - Also, expand form doesn't show all information; defeats the purpose of expanding the record
- [x] Infinite scroll on Kanban view is neat
- [ ] Rename view, page appears to reload twice
- [ ] Copy View, didn't copy `stack by` configuration
- [ ] Do not allow creating a Kanban view if there is no `select field` in the table
- [ ] Cards re-arranged with a stack: order not retained on toggling away from Kanban view; might not be feasible to implement as we do not allow reordering of records.
- [ ] Rename stack option didn't work
- [ ] Was unable to move the Uncategorised stack to any other location; by design?
- [ ] [airtable] Share view?
- [ ] [airtable] Good to display image field, if present on the card (like gallery); 
- [ ] [airtable] Only non-empty values are displayed in kanban cards; the information is also printed on customizing cards menu

**Comment by dstala** [comment]
Test scenario's
: Using sakila DB, Film table

- [x] Create kanban 
- [x] Rename kanban
- [ ] Duplicate kanban
- [x] Configure grouping field: `Rating`
  - [x] Verify stack count
  - [x] Verify stack order
  - [x] Verify footer records count
  - [x] Verify per stack card count
  - [x] Verify card order within a stack
- [x] Toolbar menu 
  - [x] Hide fields
  - [x] Enable filter
  - [x] Disable filter
  - [x] Sort fields
  - [x] Reset sort
- [x] Drag drop
  - [x] Inter stack card drag-drop
  - [x] Intra stack card drag-drop [will be reset if we re-open kanban]
  - [x] Stack drag-drop reorder
- [ ] Stack context menu
  - [ ] Add record
  - [ ] Rename stack
  - [ ] Delete stack
  - [ ] Collapse stack
- [ ] Expand record
  - [ ] Fields visible
  - [ ] Update record  
- [ ] View context menu
  - [ ] Download CSV
  - [ ] Upload CSV
  - [ ] Locked view 
- [ ] Share view
- [ ] Kanban stack handling
  - [ ] Change grouping field
  - [ ] Add / edit stack
    - [ ] Add option
    - [ ] Rename option
    - [ ] Delete option 
- [x] Delete kanban

**Review by bcakmakoglu — COMMENTED** [review]
Code wise it looks good to me, needs more investigation to discern if there's memory leaks or something like that which could cause stack buildup and slowdowns

File: packages/nc-gui/components/smartsheet/Kanban.vue [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+  // update isMounted for below watcher
+  isMounted = true
+})
+
+watch(
+  () => shouldScrollToRight.value,
+  () => {
+    // use `isMounted` to avoid scrolling from switching from other views
+    // i.e. scroll only when a new option is added within kanban view
+    if (isMounted && shouldScrollToRight.value) {
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Kanban.vue:242]: instead of using `isMounted` you could also use the watch option `{ flush: 'post' }`

File: packages/nc-gui/components/smartsheet/Kanban.vue [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+})
+
+const expandFormClick = async (e: MouseEvent, row: RowType) => {
+  if (e.target as HTMLElement) {
+    expandForm(row)
+  }
+}
+
+/** Block dragging the stack to first index (reserved for uncategorized) **/
+function onMoveCallback(event: any) {
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Kanban.vue:148]: Is there no type for this event? All of these are `any`.
Can you provide an interface?

File: packages/nc-gui/components/smartsheet/Kanban.vue [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+      })
+      // reset shouldScrollToRight
+      shouldScrollToRight.value = false
+    }
+  },
+)
+</script>
+
+<template>
+  <div class="flex h-full bg-white px-2">
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/Kanban.vue:256]: A couple of line breaks in the HTML template would help with readability.
Right now everything is just put next to each other so Hierarchies are pretty difficult to discern.

File: packages/nc-gui/components/virtual-cell/ManyToMany.vue [hunk 4]
```diff
@@ -95,20 +95,22 @@ const onAttachRecord = () => {
         <template v-if="cells">
           <ItemChip v-for="(cell, i) of cells" :key="i" :item="cell.item" :value="cell.value" @unlink="unlinkRef(cell.item)" />
 
-          <span v-if="cells?.length === 10" class="caption pointer ml-1 grey--text" @click="childListDlg = true">more... </span>
+          <span v-if="cells?.length === 10" class="caption pointer ml-1 grey--text" @click.stop="childListDlg = true"
+            >more...
+          </span>
         </template>
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/virtual-cell/ManyToMany.vue:101]: Formatting

File: packages/nc-gui/composables/useKanbanViewData.ts [hunk 5]
```diff
@@ -0,0 +1,432 @@
+import type { ComputedRef, Ref } from 'vue'
+import type { Api, ColumnType, KanbanType, SelectOptionType, SelectOptionsType, TableType, ViewType } from 'nocodb-sdk'
+import { useI18n } from 'vue-i18n'
+import { message } from 'ant-design-vue'
+import type { Row } from '~/composables/useViewData'
+import { SharedViewPasswordInj, deepCompare, enumColor, useNuxtApp } from '#imports'
+
+type GroupingFieldColOptionsType = SelectOptionType & { collapsed: boolean }
+
+export function useKanbanViewData(
```
Comment (bcakmakoglu) [inline packages/nc-gui/composables/useKanbanViewData.ts:10]: Instead of using `useState` for everything why not wrap this composable in `useInjectionState` (similar to other composables like this)
Reason:
* We execute the composable once and don't duplicate functions etc. 
* We don't create multiple injections for the same state (useState creates an injection for each value, i.e. all of these are passed through the whole tree instead of just 1 injection)

File: packages/nc-gui/composables/useSharedView.ts [hunk 6]
```diff
@@ -67,14 +77,39 @@ export function useSharedView() {
...
+        sortArrJson: JSON.stringify(sorts.value),
+        ...params,
+      } as any,
+      {
+        headers: {
+          'xc-password': password.value,
+        },
+      },
+    )
+    console.log(data)
```
Comment (bcakmakoglu) [inline packages/nc-gui/composables/useSharedView.ts:112]: remove log

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

File: packages/nc-gui/assets/css/global.css [hunk 7]
```diff
@@ -22,4 +22,11 @@ Apply Vazirmatn for rtl
...
 .rtl .v-application .ml-n1 {
 	margin-left: 0px !important;
+}
+
+/*
+For Drag and Drop
+*/
+.grabbing * {
+	cursor: grabbing;
 }
```
[the remark this answers is not in the document]
Comment (wingkwong) [reply packages/nc-gui/assets/css/global.css:32]: No such thing..

image

File: packages/nc-gui/components/smartsheet/Kanban.vue (hunk 1 above)
Comment (wingkwong) [reply packages/nc-gui/components/smartsheet/Kanban.vue:242]: Tried adding that. Not working as expected.

File: packages/nc-gui/components/smartsheet/header/CellIcon.vue [hunk 8]
```diff
@@ -85,5 +85,5 @@ const icon = computed(() => {
 </script>
 
 <template>
-  <component :is="icon" class="text-grey mx-1 !text-sm" />
+  <component :is="icon" class="text-grey mx-1 !text-xs" />
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/header/CellIcon.vue:88]: Should the icons really be reduced in size? They're already pretty small.

File: packages/nc-gui/components/smartsheet/header/VirtualCellIcon.vue [hunk 9]
```diff
@@ -77,5 +77,5 @@ const icon = computed(() => {
 </script>
 
 <template>
-  <component :is="icon.icon" class="mx-1 !text-sm" :class="icon.color" />
+  <component :is="icon.icon" class="mx-1 !text-xs" :class="icon.color" />
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/smartsheet/header/VirtualCellIcon.vue:80]: Same: Does this need to be even smaller?

File: packages/nc-gui/components/virtual-cell/Lookup.vue [hunk 10]
```diff
@@ -49,7 +49,7 @@ const lookupColumnMetaProps = useColumn(lookupColumn)
 </script>
 
 <template>
-  <div class="h-full flex gap-1">
+  <div class="h-full flex gap-1 overflow-x-auto p-1">
```
Comment (bcakmakoglu) [inline packages/nc-gui/components/virtual-cell/Lookup.vue:52]: Wouldn't this allow horizontal scrolling? Is this desired?

File: packages/nc-gui/composables/useViewData.ts [hunk 11]
```diff
@@ -66,6 +59,13 @@ export function useViewData(
 
   const { isUIAllowed } = useUIPermission()
 
+  const formatData = (list: Record<string, any>[]) =>
+    list.map((row) => ({
+      row: { ...row },
+      oldRow: { ...row },
+      rowMeta: {},
+    }))
```
Comment (bcakmakoglu) [inline packages/nc-gui/composables/useViewData.ts:67]: Why was this moved inside the composable? 
Outside it would only be created once, inside as many times as the composable is called and this function doesn't seem to have dependencies that require it to be inside

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

**Review by bcakmakoglu — CHANGES_REQUESTED** [review]

File: scripts/cypress/integration/common/00_pre_configurations.js [hunk 12]
```diff
@@ -164,7 +164,7 @@ export const genTest = (apiType, dbType) => {
 
     it("Admin SignUp", () => {
       cy.task("log", "This will be output to the terminal");
-      loginPage.signUp(roles.owner.credentials);
+      loginPage.signIn(roles.owner.credentials);
```
[the remark this answers is not in the document]
Comment (wingkwong) [reply scripts/cypress/integration/common/00_pre_configurations.js]: I think this is outdated. Not sure why we got this change. In local it's `signUp`.

File: packages/nc-gui/components/cell/SingleSelect.vue [hunk 13]
```diff
@@ -67,7 +69,7 @@ watch(isOpen, (n, _o) => {
   <a-select
     ref="aselect"
     v-model:value="vModel"
-    class="w-full"
+    class="w-full 2"
```
[the remark this answers is not in the document]
Comment (wingkwong) [reply packages/nc-gui/components/cell/SingleSelect.vue:72]: I guess I accidentally typed it. Removed.

File: packages/nc-gui/components/cell/SingleSelect.vue [hunk 14]
```diff
@@ -80,16 +82,15 @@ watch(isOpen, (n, _o) => {
...
       </a-tag>
     </a-select-option>
   </a-select>
 </template>
 
 <style scoped lang="scss">
 .rounded-tag {
-  padding: 0px 12px;
-  border-radius: 12px;
+  @apply py-0 px-[12px] rounded-[12px];
```
[the remark this answers is not in the document]
Comment (wingkwong) [reply packages/nc-gui/components/cell/SingleSelect.vue:93]: I just changed the existing style using `@apply` only.

File: packages/nc-gui/components/virtual-cell/Lookup.vue (hunk 10 above)
Comment (wingkwong) [reply packages/nc-gui/components/virtual-cell/Lookup.vue:52]: As I remember, if you got more lookup tags, they will be displayed out of the input window. So I made it overflow-x-auto atm.

File: packages/nc-gui/components/cell/Rating.vue [hunk 15]
```diff
@@ -32,7 +32,13 @@ const vModel = computed({
 </script>
 
 <template>
-  <a-rate v-model:value="vModel" :count="ratingMeta.max" :style="`color: ${ratingMeta.color}`" :disabled="!editEnabled">
+  <a-rate
+    v-model:value="vModel"
+    :count="ratingMeta.max"
+    :style="`color: ${ratingMeta.color}; padding: 0px 5px`"
+    :class="{ '!ml-[-8px]': !editEnabled }"
```
[the remark this answers is not in the document]
Comment (wingkwong) [reply packages/nc-gui/components/cell/Rating.vue:39]: 1. to align in kanban view (previously they had different spacing on the left)
2. to reduce the space in other views

image

File: packages/nc-gui/components/smartsheet/header/CellIcon.vue (hunk 8 above)
Comment (wingkwong) [reply packages/nc-gui/components/smartsheet/header/CellIcon.vue:88]: Navi wants it smaller

File: packages/nc-gui/components/smartsheet/header/VirtualCellIcon.vue (hunk 9 above)
Comment (wingkwong) [reply packages/nc-gui/components/smartsheet/header/VirtualCellIcon.vue:80]: Same as above

**Review by bcakmakoglu — APPROVED** [review]

**Comment by o1lab** [comment]
Did re-run the test few times - they are failing. @wingkwong @dstala

File: packages/nc-gui/composables/useKanbanViewStore.ts [hunk 16]
```diff
@@ -0,0 +1,548 @@
...
+    //     {row: {...}, oldRow: {...}, rowMeta: {...}},
+    //     ...
+    //   ],
+    //   [val2] : [
+    //     {row: {...}, oldRow: {...}, rowMeta: {...}},
+    //     {row: {...}, oldRow: {...}, rowMeta: {...}},
+    //     ...
+    //   ],
+    // }
+    const formattedData = ref<Record<string, Row[]>>({})
```
Comment (pranavxc) [inline packages/nc-gui/composables/useKanbanViewStore.ts:59]: You can use `Map` here, with `Map` you can have `null` as a key as well. By using map we can avoid keeping it as `uncategorized` and avoid conflict actual option with value `uncategorized`

File: packages/nc-gui/composables/useKanbanViewStore.ts [hunk 17]
```diff
@@ -0,0 +1,548 @@
...
+
+    // countByStack structure
+    // {
+    //   "uncategorized": 0,
+    //   [val1]: 10,
+    //   [val2]: 20
+    // }
+    const countByStack = ref<Record<string, number>>({})
```
Comment (pranavxc) [inline packages/nc-gui/composables/useKanbanViewStore.ts:67]: Same as above comment

**Review by bcakmakoglu — APPROVED** [review]

**Review by pranavxc — APPROVED** [review]
